src/librustc_interface/queries.rs

```diff
@@ -246,46 +253,19 @@ impl Compiler {
                     rx,
                     &*outputs.peek()
                 ))
-            })
-        })
-    }
-
-    pub fn link(&self) -> Result<&Query<()>> {
-        self.queries.link.compute(|| {
-            let sess = self.session();
+            })?;
 
-            let ongoing_codegen = self.ongoing_codegen()?.take();
+            // Drop GlobalCtxt after starting codegen to free memory.
+            mem::drop(self.global_ctxt()?.take());
 
             self.codegen_backend().join_codegen_and_link(
                 ongoing_codegen,
-                sess,
+                self.session(),
                 &*self.dep_graph()?.peek(),
-                &*self.prepare_outputs()?.peek(),
+                &*outputs.peek(),
             ).map_err(|_| ErrorReported)?;
 
             Ok(())
         })
     }
-
-    pub fn compile(&self) -> Result<()> {
```

Comment:
This is meant as an simple utility function and an example for users of `rustc_interface` of something that mirrors a regular compilation session without all the complexities of the full rustc driver (in `rustc_driver`).

Reply:
If an example is needed, why not provide it in documentation? Don't leave unnecessary code in the product.

Reply:
Do we have CI testable example code snippets in the rustc-guide?

 * Or maybe we could put such a snippet into the rustdoc comments that we process for `rustc`'s generated documentation?

If either of the above is true, then I could go along with @nnethercote 's reasoning here, though then I would want that migration to happen either before this PR, or as part of this PR.

(Otherwise, I'm with @Zoxc; Its hard enough for people to get started using the compilier as a library or hacking on `rustc` itself; simple examples of how to use things like `rustc_interface` are in their own way part of our product...)